src/test/java/com/faforever/api/user/UserServiceTest.java

```diff
@@ -162,4 +164,118 @@ public void activateTokenExpired() throws Exception {
+    verify(emailService).sendPasswordResetMail(eq("JUnit"), eq("junit@example.com"), urlCaptor.capture());
+
+    String passwordResetUrl = urlCaptor.getValue();
+    assertThat(passwordResetUrl, startsWith("http://www.example.com/resetPassword/"));
+
+    String token = passwordResetUrl.split("/")[4];
+    HashMap<String, String> claims = objectMapper.readValue(JwtHelper.decode(token).getClaims(), HashMap.class);
+
+    assertThat(claims.get(UserService.KEY_ACTION), is("reset_password"));
+    assertThat(claims.get(UserService.KEY_USER_ID), is(dummyUser().getId()));
+    assertThat(Instant.parse(claims.get(UserService.KEY_EXPIRY)).isAfter(Instant.now()), is(true));
+  }
+
+  @Test
+  @SneakyThrows
+  @SuppressWarnings("unchecked")
+  public void resetPasswordUnknownUser() {
+    expectedException.expect(ApiExceptionWithCode.apiExceptionWithCode(ErrorCode.USERNAME_INVALID));
+
+    when(userRepository.findOneByEmailIgnoreCase("junit@example.com")).thenReturn(null);
+    instance.resetPassword("junit@example.com");
+  }
+
+  @Test
+  @SneakyThrows
+  public void claimPasswordResetToken() {
+    String token = JwtHelper.encode(objectMapper.writeValueAsString(ImmutableMap.of(
+        UserService.KEY_ACTION, "reset_password",
+        UserService.KEY_EXPIRY, Instant.now().plusSeconds(100).toString(),
+        UserService.KEY_USER_ID, "5"
+    )), new MacSigner("banana")).getEncoded();
+
+    User user = dummyUser();
+
+    when(userRepository.findOne(5)).thenReturn(user);
+
+    instance.claimPasswordResetToken(token, "newPassword");
+
+    ArgumentCaptor<User> captor = ArgumentCaptor.forClass(User.class);
+    verify(userRepository).save(captor.capture());
+    assertEquals(captor.getValue().getPassword(), "5c29a959abce4eda5f0e7a4e7ea53dce4fa0f0abbe8eaa63717e2fed5f193d31");
+  }
+
+  @Test
+  @SneakyThrows
+  public void claimNonPasswordResetToken() {
+    expectedException.expect(ApiExceptionWithCode.apiExceptionWithCode(ErrorCode.TOKEN_INVALID));
+
+    String token = JwtHelper.encode(objectMapper.writeValueAsString(ImmutableMap.of(
+        UserService.KEY_ACTION, "foobar",
+        UserService.KEY_EXPIRY, Instant.now().plusSeconds(100).toString(),
+        UserService.KEY_USER_ID, "5"
+    )), new MacSigner("banana")).getEncoded();
+
+    instance.claimPasswordResetToken(token, "newPassword");
+  }
+
+  @Test
+  @SneakyThrows
+  public void claimPasswordResetTokenExpired() {
+    expectedException.expect(ApiExceptionWithCode.apiExceptionWithCode(ErrorCode.TOKEN_EXPIRED));
+
+    String token = JwtHelper.encode(objectMapper.writeValueAsString(ImmutableMap.of(
+        UserService.KEY_ACTION, "reset_password",
+        UserService.KEY_EXPIRY, Instant.now().plusSeconds(-100).toString(),
+        UserService.KEY_USER_ID, "5"
+    )), new MacSigner("banana")).getEncoded();
+
+    instance.claimPasswordResetToken(token, "newPassword");
+  }
+
+
+  private User dummyUser() {
```

Comment:
Consider creating a `Factory/Builder` for it ...https://github.com/FAForever/faf-java-api/tree/develop/src/test/java/com/faforever/integration/factories